src/NServiceBus.Core/Routing/FileBasedDynamicRouting/FileRoutingTable.cs

```diff
@@ -1,189 +1,109 @@
+
+        Dictionary<Endpoint, HashSet<EndpointInstance>> instanceMap = new Dictionary<Endpoint, HashSet<EndpointInstance>>();
+        ReadOnlySettings settings;
+        string filePath;
+        TimeSpan checkInterval;
+        IAsyncTimer timer;
+        IRoutingFileAccess fileAccess;
+        FileRoutingTableParser parser = new FileRoutingTableParser();
+        int maxLoadAttempts;
+        
+
+        public FileRoutingTable(string filePath, TimeSpan checkInterval, IAsyncTimer timer, IRoutingFileAccess fileAccess, int maxLoadAttempts, ReadOnlySettings settings)
         {
-            this.basePath = basePath;
-            this.timeToWaitBeforeRaisingFileChangedEvent = timeToWaitBeforeRaisingFileChangedEvent;
+            this.settings = settings;
+            this.filePath = filePath;
+            this.checkInterval = checkInterval;
+            this.timer = timer;
+            this.fileAccess = fileAccess;
+            this.maxLoadAttempts = maxLoadAttempts;
         }
 
-        public IEnumerable<EndpointInstance> GetInstances(Endpoint endpoint)
+        protected override async Task OnStart(IBusContext context)
         {
-            logger.DebugFormat("Request routes for {0}.", endpoint);
+            var endpointInstances = settings.Get<EndpointInstances>();
+            endpointInstances.AddDynamic(FindInstances);
 
-            CacheRoute routes;
-            if (!routeMapping.TryGetValue(endpoint.ToString(), out routes))
-            {
-                UpdateMapping(endpoint.ToString(), true);
-
-                if (!routeMapping.TryGetValue(endpoint.ToString(), out routes))
-                {
-                    yield break;
-                }
-            }
-            foreach (var route in routes.Routes)
-            {
-                var discriminators = route.Split(new []{':'},StringSplitOptions.None);
-                if (discriminators.Length == 2)
-                {
-                    var userDiscriminator = NullIfEmptyString(discriminators[0].Trim());
-                    var transportDiscriminator = NullIfEmptyString(discriminators[1].Trim());
-
-                    yield return new EndpointInstance(endpoint, userDiscriminator, transportDiscriminator);
-                }
-                else
-                {
-                    logger.Info($"Invalid route {route}. Expecting <userDiscriminator>:<transportDiscriminator> format");
-                }
-            }
-        }
+            await ReloadData();
 
-        static string NullIfEmptyString(string value)
-        {
-            return value.Equals("", StringComparison.InvariantCultureIgnoreCase) 
-                ? null 
-                : value;
+            timer.Start(ReloadData, checkInterval, ex => log.Error("Error while reading routing table", ex));
         }
 
-        void StartMonitoring(string basePath, string queueName, string fileName)
+        async Task ReloadData()
```

Comment:
Currently we have two level retries. Immediate retries done inside ReloadData and loop inside AsyncTimer do we want both? Should we enable endpoint instance to start even if we did not load a routing file a single time?